File: 1201-1300/1234-Replace the Substring for Balanced String/1234-Replace the Substring for Balanced String.py
import collections
class Solution:
    def balancedString(self, s: str) -> int:
        target = len(s) // 4
        counts = collections.Counter(s)
        start = 0
        minLen = len(s)
        for i, c in enumerate(s):
            counts[c] -= 1
            while (start < len(s) and counts['Q'] <= target and counts['W'] <= target and
                    counts['E'] <= target and counts['R'] <= target):
                minLen = min(minLen, i - start + 1)
                counts[s[start]] += 1
                start += 1

        return minLen
    

File: 1201-1300/1234-Replace the Substring for Balanced String/1234-Replace the Substring for Balanced String.cpp
class Solution {
public:
    int balancedString(string s) {
        int target = s.size() / 4;
        unordered_map<char, int> counts;
        for (char c : s) {
            ++counts[c];
        }
        int start = 0, minLen = s.size();
        for (int i = 0; i < s.size(); ++i) {
            --counts[s[i]];
            while (start < s.size() && counts['Q'] <= target &&
                counts['W'] <= target && counts['E'] <= target &&
                counts['R'] <= target) {
                minLen = min(minLen, i - start + 1);
                ++counts[s[start++]];
            }
        }

        return minLen;
    }
};
